src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+
+When creating a new service, run `npm shrinkwrap` to generate an `npm-shrinkwrap.json` based on the packages you have installed in your development environment, then check that file into the Git repository.
+When beginning development on an existing service, the usual `npm install` will install the specific versions named in the `npm-shrinkwrap.json`.
+
+*Important*: Do not use `devDependencies`, as npm versions before 3.0 do not deal with them correctly and will cause failures in deployment that do not appear in local runs or Travis.
+Instead, list all dependencies in `dependencies`.
+
+When changing a service's dependencies, the easiest approach is to use `npm install --save` to install the package and update `package.json` at the same time.
+Once this is complete, run `npm shrinkwrap` and add the modified `npm-shrinkwrap.json` to your commit.
+Pull requests with a modified `package.json` that do not have a corresponding `npm-shrinkwrap.json` diff should be viewed with suspicion!
+
+If is the service owner's responsibility to keep dependencies up to date.
+The `npm outdated` command gives a useful overview of available updates.
+In general, try to keep packages up to date within semver constraints (so, fix things displayed in red in `npm outdated`), but be cautious that the new code you are incorporating into your service is trustworthy.
+In an ideal world, that means a thorough security review.
+In the real world, that probably means a lot less.
+
+## Testing
+
+### Test Setup
+
+Use Mocha to run unit tests, in the `test/` directory.
+Include the following in `mocha.opts`, most critically the `--ui tdd`.
+
+```
+--ui tdd
+--timeout 30s
+--reporter spec
+```
+
+Name the test files `test/*_test.js`, so they will be matched by the `npm test` script given above.
+These files should require production code from `../lib/`: `foo = require('../lib/foo')`.
+
+### Helpers
+
+Include any shared test-specific code in `test/helpers.js`.
+
+### ESLint
+
+Use [eslint-config-taskcluster](https://github.com/taskcluster/eslint-config-taskcluster) and [eslint-plugin-taskcluster](https://github.com/taskcluster/eslint-plugin-taskcluster) along with `mocha-eslint` to check for lint.
+
+To do so, install `eslint`, `babel-eslint`, `mocha-eslint`, `eslint-config-taskcluster`, and `eslint-plugin-taskcluster`.
+Add the following to the `scripts` section of `package.json`: `"test": "mocha .test/lint.js .test/*_test.js"`.
+
+Create `.eslintrc` in the root of the repository with
+```js
+{
+  "extends": "eslint-config-taskcluster"
+}
+```
+
+And the following in `test/lint.js`:
+```js
+var lint = require('mocha-eslint');
+
+var paths = [
+  'src/*.js',
+  'test/*.js',
+];
+
+lint(paths);
+```
+
+### Test Requirements
+
+A simple `git clone` .. `npm install` .. `npm test` should run successfully for new contributors.
+Anything else dramatically increases the difficulty in getting started.
```

Comment:
Let's add this:

For services which require credentials, they should be possible to supply in the `user-config.yml` file.  There should be a `user-config-example.yml` file which has all the necessary settings filled with an example value or the string `'...'`.  This makes it easier for people to know which credentials they need and how to set them up.  The `.gitignore` and `.npmignore` files should exclude `user-config.yml`.